### PR TITLE
Add support for persistent walkstyles

### PR DESCRIPTION
**Pull request by crusopaul**
Currently walkstyle animations set by dpemotes will disappear when a ped is unloaded / reloaded. A common scenario where this occurs is on player logout / login - logging in does not apply the last set walkstyle.

This PR adds optional support to dpemotes for saving walkstyles that persist across player logout / login. This is achieved through the use of KVPs and a thread poll.

This was adapted from the work done for #62.

### CODE REVIEW

**Comment by crusopaul** [comment]
That's partially correct. But this PR is storing it both server and client-side.

It is correct that the way this is written could cause the server KVP to become orphaned / the player no longer able to retrieve data from the server KVP if they changed their name. However, this PR makes an initial check for the client side KVP.

Unless someones client and server KVP's both disappeared, then they would not run into the issue of a non-persisting walkstyle.

There's a dialogue I had with myself when considering the use of the name and possibility of the name changing. I have heard that return value of `GetPlayerName()` can change depending on Steam account linking and whether or not Steam is open (having trouble finding source)? I almost used the game license instead of the name, however, people may own multiple copies of the game and, thus, have different game licenses. In the case where they had a different copy of the game it is likely that they were playing it on a different device and would not have their client KVP's synced - at this point they'd probably experience a non-persisting walkstyle. Same sort of deal applies if this keyed off of the IP. It seemed like the name was the best to key off of?

If there is a better way to index the server KVP than relying on the player name then I'd definitely prefer to do that.

**Comment by AvaN0x** [comment]
> I have heard that return value of `GetPlayerName()` can change depending on Steam account linking and whether or not Steam is open (having trouble finding source)?

Yes, steam is not required for FiveM, so it is not 100% of the time a steam name.

> I almost used the game license instead of the name, however, people may own multiple copies of the game and, thus, have different game licenses.

This issue is the same for every informations then:
- rockstar license: player can have multiple gta copies
- steam: player can have multiple steam accounts
- fivem: player can have multiple cfx account
- discord: player can have multiple discord account

**But**, the one that should be used imo is the rockstar license, here are some of my thoughts about the different approach:
- fivem: I don't know a lot of player with a linked cfx account =>not always there
- steam is not required for a fivem server => not always there
- discord: not every player play with discord => not always there
- license: a rockstar account is needed to play on fivem, all player have one
- name: I don't like this one at all, random characters can be used by the user, idk how well KVP will store some names like `A̵͐͝v̷͑͘a̵̐̕N̴̑͒0̸̢͗x̶̆̄`
Also, names are not unique, I'm pretty sure you can have multiple steam accounts with the same name, same with FiveM name which is set in the settings which would lead to:
- Players using other players walkstyle (not that important)
- Players overriding other players walkstyle (well, important xD)

> people may own multiple copies of the game and, thus, have different game licenses.

If they do, they most likely already have issues saving their player data on different servers. I think you should ignore that fact and go for the Rockstar License. I have heard of the license2 thing too, I'm not sure if the best to use is `license` or `license2`.

---

TLDR: Use the license

**Comment by AvaN0x** [comment]
> Unless someones client and server KVP's both disappeared, then they would not run into the issue of a non-persisting walkstyle.

I'm not sure about the client and server side saving of this. 🤔 
If I were the one to do it, I would only have saved it server side, for user experience, I'd prefer my walkstyle to be saved on the server than on my client, so that I can play on multiple servers without having a wlakstyle that do not suit my character.
People might hate having to switch off their walking style when going to another server ? 🤔 

Just some thoughts I have, otherwise, thank you for the your PR :)

**Comment by TayMcKenzieNZ** [comment]
Interesting points, and I agree;, you could be playing as a fat man on Server 1 so your walkstyle is chubby male, but on Server 2 you're a gangsta with a gangsta walk, so having a chubby walking gangsta or a gangsta walking fat man doesn't really make any sense lol, as for people playing in the same server as multiple different characters, now that's something that's completely different and complicated and I believe that's something you leave for more complex scripts and frameworks and keep that out of DpEmotes as that just sounds like a headache, I mean some people play as six different characters, how are you supposed to store all 6 walkstyles? Don't know, might be possible, but good grief I don't even want to think about that lol

**Comment by crusopaul** [comment]
I see 😯 I'm sold on using the license for this reason alone.

> names are not unique

Even though this isn't sensitive data by any means, I wouldn't want to allow players to interact with each others data w/o it being intentional.

I think storing KVP's on both client and server is still best if that is alright? As someone who typically plays many characters on a few framework servers, I'll also be facing the same issue and to a higher degree. But even then, I'd still be benifitting from a server that utilizes persistent walkstyles in the sense that without it I would (intend, ocassionally forget) set my walkstyle on every login.

Another scenario to consider is players who play the same character on multiple servers - this playstyle would benefit most from client KVP's vs server KVP's. I have no idea what portion of players have this style as opposed to the style of playing different characters on different servers, but would say it's irrelevent because they benefit for the same reason as above.

**Comment by TayMcKenzieNZ** [comment]
I guess you could always test it out on a localhost and a VPS 🤔

**Comment by TayMcKenzieNZ** [comment]
I've had some time to think and the whole character thing sounds like something you would have to deal with on a more complex approach such as a database or framework which, makes sense why your initial pull request was qbcore focused.

You can't necessarily store "character data" unless you have something like vMenu or, whatever character selection scripts people are using.

How about this @AvaN0x @crusopaul 

So I ran everything through Kibook who helped me with my No Idle Cam. Here's her input opinion and brainstorming take on what you both have said and what I said above regarding frameworks, databases, etc:

Tay's stance on DpEmotes is that it's standalone but *supports* qbcore, therefore adding nonsense for the framework inside qbcore is backwards and should ideally be the other way around; add something to qbcore itself, to talk to and execute commands or whatever, in Dpemotes.

There's some sense to what @crusopaul is addressing; if you only have it on the server side, then go to another server as the same "character", then you'd have to set it again at least once, and if you only store it on the client, then that's problematic if you want to have a different "character" per server.

I don't think we should be focusing on "characters" though, but more so the issue at hand and the question on everyone's mind, How Do We Achieve Persistent Walkstyles?

An option to choose whether to load the walkstyle from your client or the server might be good enough. People that use the same "character" on all servers can choose to load the walkstyle from client-side, and people that use different characters on different servers can choose to have it load it from the server. 

Maybe, a little checkbox in the dpemotes menu: "Save/Load walkstyle from server" for example. 

if its on, when you join a server your walkstyle will be loaded from the one stored in the *server*. If it's off, it will use the one saved *client-side*.

It isn't exactly difficult to join a server, select your character, open DpEmotes, set walkstyle,  go about your RP, though this shouldn't be an issue with the option and suggestion stated above.

**Comment by Senlar** [comment]
One of the other common applications for persistent walk styles is when servers use resources that reload the ped model DURING a play session. Some clothing scripts would do this and certain event based script also. For example, for Halloween I have a way for players to put on "costumes" that changes their ped into a spooky one, and then they change back after a while.

**Comment by TayMcKenzieNZ** [comment]
> One of the other common applications for persistent walk styles is when servers use resources that reload the ped model DURING a play session. Some clothing scripts would do this and certain event based script also. For example, for Halloween I have a way for players to put on "costumes" that changes their ped into a spooky one, and then they change back after a while.

Ah fuck... Well... Can't really do anything about that without frameworks etc I imagine, and I want to keep dpemotes standalone but supports qbcore 😩🤔 though I'm sure that even if you reset your ped, the KVP string would kick in eventually after a small amount of time and apply, as initially my no idle cam resource was running a loop to check when the camera is active then deactivating it, rinse and repeat, but then of FiveM updated the natives and no need for loops anymore now.

**Comment by TayMcKenzieNZ** [comment]
- The current method of supporting framework in DpEmotes is the way to go (Optional, but compatible with all their weirdness)

 - Persistent walk style should be stored client side, never server-side with identifiers weirdness (and take inputs from framework only if they are present, if they are absent then the system **should** work flawlessly)

 - Keybinds could probably be moved to clientside kvp so they are compatible between servers (like fivem intended to, if no stupid people renames shit for no reason)
 
 - We should not worry about users cleaning their KVP folders, it's their shit, we are not responsible for it and we should ignore this (+ it's not the cache folder, you have to KNOW the folder to go delete it)
 
 - Walkstyle should be accessible from exports so other resources can manipulate it, with priority system for their walkstyles

- It isn't exactly difficult to select your character, select your walkstyle, , go about your RP, leave, rejoin or join another server, select your character, select your walkstyle, go about your RP, leave, rejoin or join another server, select your walkstyle, go about your RP, etc etc.

**Comment by crusopaul** [comment]
That all seems reasonable. I'll switch it to only utilize client KVP's.

To keep this PR's scope limited to "How Do We Achieve Persistent Walkstyles?", I am going to avoid addressing the following points with code changes (these seem like good candidates for future PR's):

> Keybinds could probably be moved to clientside kvp so they are compatible between servers

> Walkstyle should be accessible from exports so other resources can manipulate it, with priority system for their walkstyles

**Comment by TayMcKenzieNZ** [comment]
Sorry if I've made a huge confusing cluster fuck 😂 also feel free to join the discord, I'll rightfully give you the contributors role and then credit you her on GitHub 😇

**Comment by crusopaul** [comment]
And actually, another event is exposed `dpemotes:client:SaveWalkstyle` that will update the client KVP. The tricky part is that this requires the animation name rather than the friendly names we are used to using for /walk. Once the following is addressed by a future PR (in the sense of DP.Walks being exposed), this event can potentially be used by things like F1 menus that allow you to change walkstyles etc.

> Walkstyle should be accessible from exports so other resources can manipulate it, with priority system for their walkstyles

And yes... I have seen this in the wild with an outdated copy of DP.Walks hardcoded into the menu.

**Comment by AvaN0x** [comment]
I just thought of something, should we have a kvp name of only `walkstyle`, or should it be dpemotes bases and be something like `dpemotes_walkstyle` or `dpemotes:walkstyle` 🤔
